### PR TITLE
Handling of remote timeout

### PR DESCRIPTION
**Pull request by bombela**
Before this commit, the remote handler and the client did not agree on how to
inform the client of a remote timeout. Effectively, the client was getting an
unexpected error.

An example when this is a problem:
- A client starts a request with a timeout (a timeout is mandatory btw).
- The remote receives the requests and jumps to the remote user code handler.
- The handler reaches the deadline, and gracefully return a deadline error
  (good boy).
- There is two possible path on the client side:
  - The client reaches the deadline itself slightly before receiving the
    response from the server. The user code receives a yarpc timeout error.
  - The client receives the deadline error from the server. The user code
    should receive a timeout error form yarpc, but instead was receiving an
    unexpected remote error.

The server returns a 504 with the message "remote timeout". The client can then
build a timeout error using the remote message as the error message.

The server returns a tchannel timeout error, with a similar "remote timeout"
message. The timeout error is then built from it on the client.

It is interesting to note that a local timeout (ie, one generated by tchannel
or HTTP on the client side) will have the message "timeout".

Finally, any proxy along the way could also return a timeout error, likely with
a different message.

Programmatically the user code on the client always receives a TimeoutError,
and the error message can be used by humans for analysis.

### CODE REVIEW

**Comment by bombela** [comment]
fixes #272

File: crossdock/client/timeout/behavior.go [hunk 1]
```diff
@@ -48,15 +55,49 @@ func Run(t crossdock.T) {
...
-	fatals.Error(err, "expected a failure for timeout")
+	fatals.Error(err, "expected an error")
 
 	if transport.IsBadRequestError(err) {
 		t.Skipf("sleep/raw method not implemented: %v", err)
 		return
 	}
 
+	assert.True(transport.IsTimeoutError(err),
+		"expected a TimeoutError, got %T", err)
```
Comment (abhinav) [inline crossdock/client/timeout/behavior.go]: For crossdock, `assert` will print this message in the output for both, success and failure, so you want it to be neutral. Something like,

    assert.True(transport.IsTimeoutError(err), "returns a TimeoutError")

File: crossdock/client/timeout/behavior.go [hunk 2]
```diff
@@ -48,15 +55,49 @@ func Run(t crossdock.T) {
...
+
 	form := strings.HasPrefix(err.Error(), `timeout for procedure "sleep/raw" of service "yarpc-test" after`)
 	assert.True(form, "error message has expected prefix for timeouts, got %q", err.Error())
-	assert.True(
-		transport.IsTimeoutError(err), "error should be a TimeoutError, got %T", err)
+
+	form = strings.HasSuffix(err.Error(), `(timeout)`)
+	assert.True(form, "error message has expected suffix for timeouts, got %q", err.Error())
```
Comment (abhinav) [inline crossdock/client/timeout/behavior.go]: Same here

File: crossdock/client/timeout/behavior.go [hunk 3]
```diff
@@ -48,15 +55,49 @@ func Run(t crossdock.T) {
...
+
+	ctx, cancel := context.WithTimeout(context.Background(), 200*time.Millisecond)
+	defer cancel()
+
+	dispatcher := disp.Create(t)
+	fatals.NoError(dispatcher.Start(), "could not start Dispatcher")
+	defer dispatcher.Stop()
+
+	ch := raw.New(dispatcher.Channel("yarpc-test"))
+	_, _, err := ch.Call(yarpc.NewReqMeta(ctx).Procedure("timeoutshort/raw"), nil)
```
Comment (abhinav) [inline crossdock/client/timeout/behavior.go]: For later: We should probably combine these sleep methods on the subject to take the sleep duration as a parameter.

File: internal/errors/timeout.go [hunk 4]
```diff
@@ -37,21 +37,23 @@ type timeoutError struct {
...
 	}
 }
 
 func (e timeoutError) timeoutError() {}
 
 func (e timeoutError) Error() string {
-	return fmt.Sprintf(`timeout for procedure %q of service %q after %v`,
-		e.Procedure, e.Service, e.Duration)
+	return fmt.Sprintf(`timeout for procedure %q of service %q after %v (%v)`,
+		e.Procedure, e.Service, e.Duration, e.Message)
```
Comment (abhinav) [inline internal/errors/timeout.go]: Instead of `(%v)`, it may be preferable to use `: %v`

File: transport/http/handler.go [hunk 5]
```diff
@@ -59,6 +59,11 @@ func (h handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		return
 	}
 
+	if err == context.DeadlineExceeded {
+		http.Error(w, "remote timeout", http.StatusGatewayTimeout)
```
Comment (abhinav) [inline transport/http/handler.go]: I think we should include all the available information in this error:

    timeout for procedure %q of service %q from caller %q after %v

Then on the client side, for remote timeouts, instead of using the same `NewTimeoutError`, you'll use the 'remote' version which only accepts the full message, which is assumed to have all the necessary information.

Comment (abhinav) [inline transport/http/handler.go]: The advantage of doing it this way is that you get better, readable error messages even when the client is not a YARPC client.

Comment (bombela) [inline transport/http/handler.go]: there is no local vs remote version of the timeout error in tchannel. So I decided to not add the distinction in yarpc. So really, there is 3 kind of timeout error: local, remote and "local_or_remote_i_cant_tell". With HTTP the distinction is clear (the http transport returns deadlines errors locally, and we have the 504 for a remote returning a timeout). TChannel has a higher abstraction, returning the same timeout error code for local timeout and remote timeout.

The distinction is mostly for human consumption and for a couple tests in crossdock. I decided to use a single timeout error, with an msg containing extra details (and I am all for adding more details as you proposed).

Comment (bombela) [inline transport/http/handler.go]: After many discussion in person, we decided to have different types of timeout behind the TimeoutError interface:

 - handlerTimeoutError for representing a timeout error on the server side. It includes caller, service, procedure and the time waited.
 - remoteTimeoutError for the client representation of a handlerTimeoutError.
 - clientTimeoutError, for representing a timeout error on the client side, it includes the service, procedure and time waited.

Because TChannel returns to the client the same error code in both client side and server side timeouts, all timeout errors via TChannel will be represented as a remoteTimeoutError. Overtime, we can work to improve the situation in TChannel (at least more informative error messages).

**Comment by bombela** [comment]
here we go, all case should now be tested on http & tchannel:
 - yarpc client, client side timeout
 - yarpc client, remote handler side timeout
 - yarpc server, handler timeout

File: crossdock/client/errors/behavior.go [hunk 6]
```diff
@@ -311,11 +312,29 @@ func Run(t crossdock.T) {
...
+			wantBodyStartsWith: `handler timeout for procedure "waitfortimeout/raw"` +
+				` of service "yarpc-test" from caller "yarpc-test" after`,
+		},
 	}
 
 	for _, tt := range tests {
 		res := client.Call(t, tt.headers, tt.body)
 		t.Tag("scenario", tt.name)
+		if res.Status == tt.skipStatus {
+			t.Skipf("waitfortimeout/raw method not implemented: %v", res.Body)
```
Comment (abhinav) [inline crossdock/client/errors/behavior.go]: Don't hard code the method name ("waitfortimeout/raw"). Just use a generic message with `tt.name`. That will give us enough information about what got skipped and why.

File: crossdock/client/errorstchout/behavior.go [hunk 7]
```diff
@@ -150,6 +152,27 @@ func Run(t crossdock.T) {
...
 			},
 		},
+		{
+			name:      "remote timeout error",
+			procedure: "waitfortimeout/raw",
+			body:      []byte("{}"),
+			headers:   []byte{0x0, 0x0},
+			encoding:  "raw",
+			validate: func(res3 []byte, isAppErr bool, err error) {
+				assert.Error(err, "is error")
```
Comment (abhinav) [inline crossdock/client/errorstchout/behavior.go]: nit: Maybe use "request must fail"

Also, if this returns false, there's no point in doing any of the other checks so return right away:

```go
if !assert.Error(err, "...") {
    return
}
```

File: crossdock/client/errorstchout/behavior.go [hunk 8]
```diff
@@ -150,6 +152,27 @@ func Run(t crossdock.T) {
...
+				assert.False(isAppErr, "waitfortimeout/raw procedure must not produce application error")
```
Comment (abhinav) [inline crossdock/client/errorstchout/behavior.go]: Same here: If this returns false, then the cast on the next line will panic so short circuit early. 

```go
if !assert.False(...) { return }
```

File: crossdock/client/httpserver/behavior.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+	fatals.Error(err, "expected an error")
+
+	if transport.IsBadRequestError(err) {
+		t.Skipf("handlertimeout/raw method not implemented: %v", err)
+		return
+	}
+
+	assert.True(transport.IsTimeoutError(err), "returns a TimeoutError: %T", err)
+
+	form := strings.HasPrefix(err.Error(),
```
Comment (abhinav) [inline crossdock/client/httpserver/behavior.go]: "form"?

File: crossdock/client/tchserver/raw.go [hunk 10]
```diff
@@ -21,25 +21,32 @@
...
 	"github.com/crossdock/crossdock-go"
 	"golang.org/x/net/context"
 )
 
 func runRaw(t crossdock.T, dispatcher yarpc.Dispatcher) {
+	hello(t, dispatcher)
+	remoteTimeout(t, dispatcher)
+}
+
+func hello(t crossdock.T, dispatcher yarpc.Dispatcher) {
```
Comment (abhinav) [inline crossdock/client/tchserver/raw.go]: rename to `localTimeout`?

File: crossdock/client/tchserver/raw.go [hunk 11]
```diff
@@ -49,13 +56,42 @@ func runRaw(t crossdock.T, dispatcher yarpc.Dispatcher) {
...
+	fatals := crossdock.Fatals(t)
+
+	headers := yarpc.NewHeaders()
+	token := random.Bytes(5)
+
+	_, _, err := rawCall(dispatcher, headers, "handlertimeout/raw", token)
+	if skipOnConnRefused(t, err) {
+		return
+	}
+	fatals.Error(err, "expected an error")
```
Comment (abhinav) [inline crossdock/client/tchserver/raw.go]: Don't you want to assert that you got an error before you skip based on its value?

File: crossdock/server/tch/timeout.go [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+package tch
+
+import (
+	"time"
+
+	"github.com/uber/tchannel-go"
+	"github.com/uber/tchannel-go/raw"
+	"github.com/yarpc/yarpc-go/internal/errors"
+	"golang.org/x/net/context"
```
Comment (abhinav) [inline crossdock/server/tch/timeout.go]: import group ordering

File: internal/errors/timeout.go [hunk 13]
```diff
@@ -33,25 +33,78 @@ type TimeoutError interface {
...
-// service, procedure, and duration waited.
-func NewTimeoutError(Service string, Procedure string, Duration time.Duration) TimeoutError {
-	return timeoutError{
+var _ TimeoutError = handlerTimeoutError{}
+var _ HandlerError = handlerTimeoutError{}
+
+// HandlerTimeoutError constructs an instance of a TimeoutError representing
+// a timeout that occurred during the handler execution, with the caller,
+// service, procedure and duration waited.
+func HandlerTimeoutError(Caller string, Service string, Procedure string, Duration time.Duration) TimeoutError {
```
Comment (abhinav) [inline internal/errors/timeout.go]: As discussed offline, return `error` here

File: crossdock/server/http/server.go [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+)
+
+var server *http.Server
+var serveMux *yarpcHTTPMux
```
Comment (abhinav) [inline crossdock/server/http/server.go]: Why globals?

File: internal/errors/timeout.go [hunk 15]
```diff
@@ -33,25 +33,78 @@ type TimeoutError interface {
...
+		Procedure: Procedure,
+		Duration:  Duration,
+	}
+}
+
+func (handlerTimeoutError) timeoutError() {}
+func (handlerTimeoutError) handlerError() {}
+
+func (e handlerTimeoutError) Error() string {
+	return fmt.Sprintf(`handler timeout for procedure %q of service %q from caller %q after %v`,
```
Comment (abhinav) [inline internal/errors/timeout.go]: Change to, "call to procedure %q of service %q from caller %q timed out after %v"

File: internal/errors/timeout.go [hunk 16]
```diff
@@ -33,25 +33,78 @@ type TimeoutError interface {
...
+}
+
+type remoteTimeoutError string
+
+var _ TimeoutError = remoteTimeoutError("")
+
+// RemoteTimeoutError builds a new TimeoutError with the given message.
+//
+// It represents a TimeoutError from a remote handler.
+func RemoteTimeoutError(message string) TimeoutError {
```
Comment (abhinav) [inline internal/errors/timeout.go]: FYI if your usage is,

    type foo Bar

    func Foo(b Bar) { return foo(b) }

You can do,

    type Foo Bar

Usage will still be `Foo(bar)`.

And since this is an internal API, we don't have to worry about users relying on it always being an alias for `string`.

File: internal/errors/transport.go [hunk 17]
```diff
@@ -90,7 +102,7 @@ func (e UnrecognizedProcedureError) Error() string {
 
 // AsHandlerError for UnrecognizedProcedureError.
 func (e UnrecognizedProcedureError) AsHandlerError() HandlerError {
-	return LocalBadRequestError(e)
+	return HandlerBadRequestError(e)
```
Comment (abhinav) [inline internal/errors/transport.go]: Can we move these renames into a different PR?

File: transport/http/handler.go [hunk 18]
```diff
@@ -93,8 +96,20 @@ func (h handler) callHandler(w http.ResponseWriter, req *http.Request) error {
...
 
+	start := time.Now()
+
 	// TODO capture and handle panic
-	return h.Handler.Handle(ctx, httpOptions, treq, newResponseWriter(w))
+	err = h.Handler.Handle(ctx, httpOptions, treq, newResponseWriter(w))
+
+	// The handler is well behaved and stopped work on context deadline. We
+	// forward this information to the client diligently.
+	if err != nil && err == context.DeadlineExceeded && err == ctx.Err() {
```
Comment (abhinav) [inline transport/http/handler.go]: Nit: Don't need to check `!= nil` if you're already checking `== context.DeadlineExceeded`

File: crossdock/client/errors/behavior.go (hunk 6 above)
Comment (bombela) [inline crossdock/client/errors/behavior.go]: copy paste fail

File: crossdock/client/httpserver/behavior.go (hunk 9 above)
Comment (bombela) [inline crossdock/client/httpserver/behavior.go]: for some reason, existing code was using "form". So I used it, in lack of better idea.

File: crossdock/client/tchserver/raw.go (hunk 11 above)
Comment (bombela) [inline crossdock/client/tchserver/raw.go]: yep

File: crossdock/server/http/server.go (hunk 14 above)
Comment (bombela) [inline crossdock/server/http/server.go]: originally because I was thinking implementing Stop(). It also mirrors the tchannel makeshift server.

File: transport/http/handler.go (hunk 18 above)
Comment (bombela) [inline transport/http/handler.go]: https://golang.org/ref/spec#Comparison_operators

> Interface values are comparable. Two interface values are equal if they have identical dynamic types and equal dynamic values or if both have value nil.

So != nil kinda shorcuts the costly interface comparaison?

Ok ok, Probably not costly, err != nil removed.

File: internal/errors/transport.go (hunk 17 above)
Comment (bombela) [inline internal/errors/transport.go]: As discussed, its doable, but time consuming at this point.

File: crossdock/client/httpserver/behavior.go [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+	"strings"
+	"time"
+
+	crossdock "github.com/crossdock/crossdock-go"
+	yarpc "github.com/yarpc/yarpc-go"
+	"github.com/yarpc/yarpc-go/crossdock/client/params"
+	"github.com/yarpc/yarpc-go/encoding/raw"
+	"github.com/yarpc/yarpc-go/transport"
+	ht "github.com/yarpc/yarpc-go/transport/http"
+	"golang.org/x/net/context"
```
Comment (abhinav) [inline crossdock/client/httpserver/behavior.go]: import group ordering

File: crossdock/server/http/server.go (hunk 14 above)
Comment (abhinav) [inline crossdock/server/http/server.go]: Nit: It looks like you don't need the globals anymore. I would move them into `Start`.

File: internal/errors/timeout.go [hunk 20]
```diff
@@ -33,25 +33,72 @@ type TimeoutError interface {
...
+		Procedure: Procedure,
+		Duration:  Duration,
+	}
+}
+
+func (handlerTimeoutError) timeoutError() {}
+func (handlerTimeoutError) handlerError() {}
+
+func (e handlerTimeoutError) Error() string {
+	return fmt.Sprintf(`call to procedure %q of service %q from caller %q timed out after %v`,
```
Comment (abhinav) [inline internal/errors/timeout.go]: I would add a "Timeout: " prefix here similar to how Unexpected and BadRequest do it.

File: internal/errors/timeout.go [hunk 21]
```diff
@@ -33,25 +33,72 @@ type TimeoutError interface {
...
+
+// RemoteTimeoutError represents a TimeoutError from a remote handler.
+type RemoteTimeoutError string
+
+var _ TimeoutError = RemoteTimeoutError("")
+
+func (RemoteTimeoutError) timeoutError() {}
+
+func (e RemoteTimeoutError) Error() string {
+	return fmt.Sprintf(`remote timeout: %s`, string(e))
```
Comment (abhinav) [inline internal/errors/timeout.go]: Once the handler is adding the prefix, this can probably just be the string as-is.

File: internal/errors/timeout.go [hunk 22]
```diff
@@ -33,25 +33,72 @@ type TimeoutError interface {
...
 }
 
-func (e timeoutError) timeoutError() {}
+func (clientTimeoutError) timeoutError() {}
+func (clientTimeoutError) clientError()  {}
 
-func (e timeoutError) Error() string {
-	return fmt.Sprintf(`timeout for procedure %q of service %q after %v`,
+func (e clientTimeoutError) Error() string {
+	return fmt.Sprintf(`client timeout for procedure %q of service %q after %v`,
```
Comment (abhinav) [inline internal/errors/timeout.go]: maybe change this to

    Client Timeout: call to procedure %q of service %q timed out after %v

**Comment by abhinav** [comment]
This is great! Couple nits but LGTM otherwise.

File: internal/errors/timeout.go (hunk 20 above)
Comment (bombela) [inline internal/errors/timeout.go]: Ok, note that this error can only be created on the handler side. It is then prefixed by "remote timeout:" on the client side. So it will look like "remote timeout: Timeout: blabla" now.

File: crossdock/client/errors/behavior.go [hunk 23]
```diff
@@ -311,11 +313,60 @@ func Run(t crossdock.T) {
...
+				"RPC-Encoding":   "raw",
+				"Context-TTL-MS": "100",
+			},
+			wantStatus: 504,
+			skipStatus: 400,
+			wantBodyStartsWith: `call to procedure "waitfortimeout/raw"` +
+				` of service "yarpc-test" from caller "yarpc-test" timed out after`,
+		},
+		{
+			// We call sleep trough the proxy Phone.
```
Comment (breerly) [inline crossdock/client/errors/behavior.go]: through*

File: crossdock/server/yarpc/server.go [hunk 24]
```diff
@@ -85,6 +85,8 @@ func register(reg transport.Registry) {
 	json.Register(reg, json.Procedure("unexpected-error", UnexpectedError))
 	json.Register(reg, json.Procedure("bad-response", BadResponse))
 	json.Register(reg, json.Procedure("phone", Phone))
+	json.Register(reg, json.Procedure("sleep", Sleep))
 
 	raw.Register(reg, raw.Procedure("sleep/raw", SleepRaw))
+	raw.Register(reg, raw.Procedure("waitfortimeout/raw", WaitForTimeoutRaw))
```
Comment (breerly) [inline crossdock/server/yarpc/server.go]: Why have the term `raw` in the procedure name? Why not just `waitfortimeout`?

File: docker-compose.yml [hunk 25]
```diff
@@ -36,11 +37,12 @@ services:
             - BEHAVIOR_TCHCLIENT=client,server,encoding
             - BEHAVIOR_TCHSERVER=client,server,encoding
             # BEHAVIOR_HTTPCLIENT TODO
-            # BEHAVIOR_HTTPSERVER TODO
+            - BEHAVIOR_HTTPSERVER=client,httpserver
```
Comment (breerly) [inline docker-compose.yml]: I don't understand the name of this behavior? What are we testing here exactly...?

**Comment by breerly** [comment]
lgtm

File: docker-compose.yml (hunk 25 above)
Comment (bombela) [inline docker-compose.yml]: I used the confusing existing convention. This tests yarpc-client with http transport against an http server. I have mentioned few time how our tests are confusingly named, I intend to go through a renaming pass at some point.

File: crossdock/server/yarpc/server.go (hunk 24 above)
Comment (bombela) [inline crossdock/server/yarpc/server.go]: Just following the sleep/raw convention...

**Comment by blampe** [comment]
Woohoo!
